File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getHeight(self, root):
        if not root:
            return 0
        
        left = self.getHeight(root.left)
        right = self.getHeight(root.right)
        
        return max(left, right) + 1
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        left = self.getHeight(root.left)
        right = self.getHeight(root.right)
        
        return abs(right-left) <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right)

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

function getHeight(root) {
    if (root === null) {
        return -1;
    }
    return 1 + Math.max(getHeight(root.left), getHeight(root.right));
}

var isBalanced = function(root) {
    if (root == null) {
        return true;
    }
    
//     let diff = Math.abs(getHeight(root.right) - getHeight(root.left));
    
//     return diff <= 1 && isBalanced(root.left) && isBalanced(root.right);
    
    let diff = Math.abs(getHeight(root.left) - getHeight(root.right));
    return diff < 2
            && isBalanced(root.left)
            && isBalanced(root.right);
};
